vm/systemSmartContracts/liquidStaking.go

```diff
@@ -0,0 +1,244 @@
+		return nil, vm.ErrNilHasher
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, vm.ErrNilEpochNotifier
+	}
+
+	l := &liquidStaking{
+		eei:                      args.Eei,
+		delegationMgrSCAddress:   args.DelegationMgrSCAddress,
+		endOfEpochAddr:           args.EndOfEpochAddress,
+		liquidStakingSCAddress:   args.LiquidStakingSCAddress,
+		gasCost:                  args.GasCost,
+		marshalizer:              args.Marshalizer,
+		hasher:                   args.Hasher,
+		liquidStakingEnableEpoch: args.EpochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
+	}
+	log.Debug("liquid staking: enable epoch", "epoch", l.liquidStakingEnableEpoch)
+
+	args.EpochNotifier.RegisterNotifyHandler(l)
+
+	return l, nil
+}
+
+// Execute calls one of the functions from the delegation contract and runs the code according to the input
+func (l *liquidStaking) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	l.mutExecution.RLock()
+	defer l.mutExecution.RUnlock()
+
+	err := CheckIfNil(args)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !l.flagLiquidStaking.IsSet() {
+		l.eei.AddReturnMessage("liquid staking contract is not enabled")
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case core.SCDeployInitFunctionName:
+		return l.init(args)
+	case "claimDelegatedPosition":
+		return l.claimDelegatedPosition(args)
+	case "claimRewardsFromPosition":
+		return l.claimRewardsFromDelegatedPosition(args)
+	case "reDelegateRewardsFromPosition":
+		return l.reDelegateRewardsFromPosition(args)
+	case "unDelegateWithPosition":
+		return l.unDelegateWithPosition(args)
+	case "returnPosition":
+		return l.returnPosition(args)
+	}
+
+	l.eei.AddReturnMessage(args.Function + " is an unknown function")
+	return vmcommon.UserError
+}
+
+func (l *liquidStaking) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, l.liquidStakingSCAddress) {
+		l.eei.AddReturnMessage("invalid caller")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("not a payable function")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 1 {
+		l.eei.AddReturnMessage("invalid number of arguments")
+		return vmcommon.UserError
+	}
+	tokenID := args.Arguments[0]
+	l.eei.SetStorage([]byte(tokenIDKey), tokenID)
+
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) getTokenID() []byte {
+	return l.eei.GetStorage([]byte(tokenIDKey))
+}
+
+func (l *liquidStaking) checkArgumentsWhenPositionIsInput(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.ESDTTransfers) < 1 {
+		l.eei.AddReturnMessage("function requires liquid staking input")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("function is not payable in eGLD")
+		return vmcommon.UserError
+	}
+	for _, esdtTransfer := range args.ESDTTransfers {
+		if !bytes.Equal(esdtTransfer.ESDTTokenName, l.getTokenID()) {
```

Comment:
tokenID := l.getTokenID() outside this for and then you can use directly tokenID here, instead calling l.getTokenID() at each iteration

Reply:
done in next PR

Reply:
ok